torch/csrc/jit/mobile/function.cpp

```diff
@@ -197,6 +200,34 @@ const std::vector<int64_t>& Function::getExceptionDebugHandles() const {
   return getInterpretersExceptionDebugHandles();
 }
 
+Function& Function::registerFunc(
+    const std::string qualified_name,
+    const std::vector<Instruction>& instructions,
+    const std::vector<OperatorString>& operators,
+    const std::vector<c10::IValue> constants,
+    const std::vector<c10::TypePtr> types,
+    const google::int64 register_size) {
+  static Function func = Function(c10::QualifiedName(qualified_name));
+  for (auto const& inst : instructions) {
```

Comment:
I think this line an below would still be executed even with static in line 210.

Reply:
Yeah that's right. I was thinking that the instruction and etc will still be initialized here?

Reply:
If it's initialized here, it should be initialized only once. Although func is static, it's instructions, constants, types, etc, are appended each time this function is called. If there are multiple upgraders, this func would have all instructions from different upgraders. A simple example: 
```C++
#include <iostream>
#include <string>
#include <vector>

std::string& GetAString(int i) {
    static std::string s;
    s += std::to_string(i);
    return s;
}


int main()
{
    std::vector<std::string> vs;
    for (int i = 0; i < 3; ++i) {
        vs.emplace_back(GetAString(i));
    }
    
    for (const auto& s : vs) {
        std::cout << s << " ";
    }

    return 0;
}
```
output:
```
0 01 012 
```

So the entire upgrader table needs to be static, like @zhxchen17 suggested in another comment. If the same upgrader appears the second time, we check if the name is already in the registry, to avoid appending the elements again.

Reply:
Hmm the entire upgrade table is static, in `upgrader_mobile.h`. It's like following and this function will only be called once.

```
static std::vector<ByteCodeFunctionWithOperator> kUpgraderByteCode(
    {ByteCodeFunctionWithOperator({
        mobile::Function::registerFunc(
            "div_Tensor_0_3",
            std::vector<Instruction>({
                Instruction{OpCode::STOREN, 1, 2},
                Instruction{OpCode::LOAD, 1, 0},
                Instruction{OpCode::OP, 0, 0},
                ****
                Instruction{OpCode::RET, 0, 0},
            }), // instructions_
            std::vector<c10::IValue>({
                c10::IValue("trunc"),
                c10::IValue(true),
            }), // constants
            std::vector<c10::TypePtr>(), // types
            4 // register_size_
            ),
        std::vector<OperatorString>({
            OperatorString({"aten::is_floating_point", "", 1}),
            OperatorString({"aten::div", "Tensor", 2}),
            OperatorString({"aten::div", "Tensor_mode", 3}),
        }), // op_names

    })});
```
Do I miss anything?

Reply:
Address in latest commit.